tcptracer-bpf.c

```diff
@@ -864,17 +853,94 @@ int kretprobe__inet_csk_accept(struct pt_regs *ctx)
 		return 0;
 	}
 
+    struct fd_info t = { .active = 1 };
+    t.start_time_ns = bpf_ktime_get_ns();
+
+    bpf_map_update_elem(&active_fds, &newsk, &t, BPF_ANY);
+
 	return assert_tcp_record(newsk, status, DIRECTION_INCOMING, STATE_INITIALIZING);
 }
 
+// Is kretprobe__inet_csk_accept sufficient or is this one needed as well or instead?
+
+// SEC("kprobe/__x64_sys_accept4")
+// int kprobe__x64_sys_accept4(struct pt_regs *ctx)
+// {
+	// The file descriptor is the value returned from the syscall.
+	// int fd = PT_REGS_RC(ctx);
+	// bool t = true;
+	// bpf_map_update_elem(&active_fds, &fd, &t, BPF_ANY);
+
+// 	return 0;
+// }
+
+SEC("kprobe/__x64_sys_writev")
+int kprobe__sys_writev(struct pt_regs *ctx)
+{
+    int fd = PT_REGS_PARM1(ctx);
+    struct iovec *vectors = (struct iovec *)PT_REGS_PARM2(ctx);
+
+    struct fd_info *res = bpf_map_lookup_elem(&active_fds, &fd);
+    if (res == NULL) {
+//		 This file descriptor is not for a socket we're tracking
+        return 0;
+    }
+
+    int zero = 0;
+    void *data = bpf_map_lookup_elem(&write_buffer_heap, &zero);
+    if (data == NULL) {
+        return 0;
+    }
+
+    struct iovec vec = {};
+    bpf_probe_read(&vec, sizeof(struct iovec), vectors);
+	return 0;
+}
+
+
 SEC("kprobe/tcp_sendmsg")
 int kprobe__tcp_sendmsg(struct pt_regs *ctx) {
```

Comment:
I notice there is also a `tcp_sendpage`. There is no response tracking there. Do we know for sure that it is not needed there? A quick search doesn't show me exactly how the interaction is different.

Reply:
According to the diagram of linux tcp functions (http://vger.kernel.org/~davem/tcp_output.html) we should indeed hook there well